backend_uml.go

```diff
@@ -0,0 +1,200 @@
+// +build linux
+// +build amd64
+
+package fakemachine
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+)
+
+type umlBackend struct {
+	machine         Machine
+	kernelPath      string
+	moduleDir       string
+	slirpHelperPath string
+}
+
+func newUmlBackend(m Machine) umlBackend {
+	b := umlBackend{}
+	b.machine = m
+
+	b.kernelPath = "/usr/bin/linux.uml"
+	b.moduleDir = "/usr/lib/uml/modules"
+	b.slirpHelperPath = "/usr/bin/slirp-helper"
+
+	// TODO remove chris debug paths
+	base := "/home/obbardc/projects/debos/uml/"
+	b.kernelPath = base + "linux-uml/linux"
+	b.moduleDir = base + "linux-uml/modules/lib/modules"
+	b.slirpHelperPath = base + "libslirp-rs/target/debug/libslirp-helper"
+	return b
+}
+
+func (b umlBackend) Name() string {
+	return "uml"
+}
+
+func (b umlBackend) Supported() (bool, error) {
+	// check the binaries are present
+	if _, err := os.Stat(b.kernelPath); err != nil {
+		return false, fmt.Errorf("user-mode-linux not installed")
+	}
+	if _, err := os.Stat(b.moduleDir); err != nil {
+		return false, fmt.Errorf("user-mode-linux not installed")
+	}
+	if _, err := os.Stat(b.slirpHelperPath); err != nil {
+		// TODO confirm package name
+		return false, fmt.Errorf("slirp-helper not installed")
+	}
+	return true, nil
+}
+
+func (b umlBackend) KernelPath() (string, string, error) {
+	// check the required paths exist
+	if _, err := b.Supported(); err != nil {
+		return "", "", err
+	}
+
+	return b.kernelPath, b.moduleDir, nil
+}
+
+func (b umlBackend) InitrdModules() []string {
+	return []string{}
+}
+
+func (b umlBackend) UdevRulesImageMatch() string {
+	return "ubd*"
+}
+
+func (b umlBackend) NetworkdMatch() string {
+	return "vec*"
+}
+
+func (b umlBackend) JobOutputTTY() string {
+	// Send the fakemachine job output to the right console
+	if b.machine.showBoot {
+		return "/dev/tty0"
+	}
+	return "/dev/tty1"
+}
+
+func (b umlBackend) MountParameters(mount mountPoint) (fstype string, options []string) {
+	fstype = "hostfs"
+	options = []string{mount.hostDirectory}
+	return
+}
+
+func (b umlBackend) InitModules() []string {
+	return []string{}
+}
+
+func (b umlBackend) InitExtraVolumes() []mountPoint {
+	// mount the UML modules over the top of /lib/modules
+	// which contains the modules from the base system
+	moduleVolume := mountPoint{hostDirectory: b.moduleDir, machineDirectory: "/lib/modules", label: "modules", static: true}
+	return []mountPoint{moduleVolume}
+}
+
+func (b umlBackend) Start() (int, error) {
+	// TODO commandWrapper forces network-wait-online & doesnt exit after network setup fail ?
+	// TODO disable lvm2-activation-generator ?
+	// TODO linux process is leftover when manually killing fakemachine. this bug also happens on KVM so should be tackled in follow-up
+	// TODO killing helper doesn't stop fakemachine
```

Comment:
can you add issues for things that are orthogonal to these PR's ? :)

Reply:
sure will ;-)